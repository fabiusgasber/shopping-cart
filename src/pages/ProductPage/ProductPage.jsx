import { useLocation, useOutletContext } from "react-router-dom";
import { useState } from "react";
import ProductCard from "../../components/ProductCard/ProductCard";
import styles from "./ProductPage.module.css"
import ErrorPage from "../ErrorPage/ErrorPage";

const ProductPage = ({ propItem }) => {
    const location =  useLocation()?.state?.item;
    const item = propItem ? propItem : location;
    const addToBag = useOutletContext();
    const [quantity, setQuantity] = useState(1);
    
    if(!item) return <ErrorPage />

    return (
        <>
        <main className={styles.main} aria-live="polite">
        <ProductCard style={styles.figure} url={item.image} alt="product" />
        <section className={styles.section}>
            <article className={styles.productInfo}>
                <div id={styles.itemHeader}>
                <h1>{item.title}</h1>
                <p id={styles.price}>{item.price} $</p>
                </div>
                <p>{item.description}</p>
            </article>
            <form onSubmit={(e) => addToBag(e, item, quantity)} action="" method="get" className={styles.buttonArea}>
                <div id={styles.quantityBtn}>
                    <button name="decrease" aria-label="decrease quantity" id="decrease-btn" type="button" onClick={() => setQuantity((quantity) => quantity > 1 ? quantity - 1 : quantity)}>-</button>
                    <output name="quantity" htmlFor="decrease-btn increase-btn">{quantity}</output>
                    <button name="increase" aria-label="increase quantity" id="increase-btn" type="button" onClick={() => setQuantity((quantity) => quantity < 10 ? quantity + 1 : quantity)}>+</button>
                </div>
                <button type="submit" className="primaryBtn">Add to bag</button>
            </form>
        </section>
        </main>
        </>
    )
}

export default ProductPage;