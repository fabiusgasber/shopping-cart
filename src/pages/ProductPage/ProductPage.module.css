.main {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
  column-gap: 5rem;
}

.section {
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  gap: 2rem;
}

.section h1 {
  font-size: 2rem;
  font-family: bebas;
  font-weight: 400;
}

.productInfo {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

#itemHeader {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 1rem;
}

#price {
  white-space: nowrap;
}

.buttonArea {
  display: flex;
  gap: 2rem;
  align-items: center;
}

#quantityBtn {
  display: flex;
  gap: 1.5rem;
  align-items: center;
  font-size: 1.2rem;
}

.figure {
  display: flex;
  justify-content: center;
}

@media (max-width: 720px) {
  .main {
    justify-content: center;
  }
  .buttonArea {
    justify-content: center;
  }
}
