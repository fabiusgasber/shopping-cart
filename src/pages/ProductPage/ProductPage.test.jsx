import { it, describe, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import ProductPage from "./ProductPage";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

describe("Product Page", () => {
  const dummyItem = {
    title: "Test Item",
    price: 20.99,
    description: "Lorem ipsum",
  };

  it("renders product page", () => {
    const { container } = render(
      <MemoryRouter>
        <ProductPage />
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders a product", () => {
    render(
      <MemoryRouter>
        <ProductPage propItem={dummyItem} />
      </MemoryRouter>,
    );
    const title = screen.getByRole("heading", { name: dummyItem.title });
    const descr = screen.getByText(dummyItem.description);
    const price = screen.getByText(dummyItem.price + " $");
    expect(title).toBeInTheDocument();
    expect(descr).toBeInTheDocument();
    expect(price).toBeInTheDocument();
  });

  it("renders default value for quantity", () => {
    render(
      <MemoryRouter>
        <ProductPage propItem={dummyItem} />
      </MemoryRouter>,
    );
    const quantityOutput = screen.getByRole("status");
    expect(parseInt(quantityOutput.textContent)).toBe(1);
  });

  it("lets users select a quantity", async () => {
    const user = userEvent.setup();

    render(
      <MemoryRouter>
        <ProductPage propItem={dummyItem} />
      </MemoryRouter>,
    );

    const decreaseBtn = screen.getByRole("button", {
      name: "decrease quantity",
    });
    const increaseBtn = screen.getByRole("button", {
      name: "increase quantity",
    });
    const quantityOutput = screen.getByRole("status");

    await user.click(increaseBtn);
    await user.click(increaseBtn);
    await user.click(decreaseBtn);

    expect(parseInt(quantityOutput.textContent)).toBe(2);
  });

  it.todo(
    "should call the addToBag function when user clicks add to bag",
    async () => {
      const user = userEvent.setup();
      const addToBag = vi.fn();
      render(
        <MemoryRouter>
          <ProductPage propItem={dummyItem} onClick={addToBag} />
        </MemoryRouter>,
      );
      const addToBagBtn = screen.getByRole("button", { name: "Add to bag" });

      await user.click(addToBagBtn);

      expect(addToBag).toHaveBeenCalled();
    },
  );
});
