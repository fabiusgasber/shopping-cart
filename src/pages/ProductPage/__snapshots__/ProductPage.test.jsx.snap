// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Page > renders product page 1`] = `
<div>
  <main
    class="_main_6adc08"
  >
    <img
      src="/error.png"
    />
    <h1>
      Oh no, this page doesn't exist!
    </h1>
    <p>
      You can go back to the home page by clicking 
      <a
        data-discover="true"
        href="/"
      >
        here
      </a>
      !
    </p>
  </main>
</div>
`;
